### PR TITLE
refactor: loadUsers and move to separate file

### PR DESCRIPTION
**Pull request by Udit-takkar**
## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [x] N/A I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

### CODE REVIEW

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+export const loadUsers = async (
+  eventType: NewBookingEventType,
+  dynamicUserList: string[],
+  req: IncomingMessage
+) => {
+  try {
+    return eventType.id
+      ? await loadUsersByEventType(eventType)
+      : await loadDynamicUsers(dynamicUserList, req);
```
Comment (Udit-takkar) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:24]: separated the logic in these two functions

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  if (!Array.isArray(dynamicUserList) || dynamicUserList.length === 0) {
+    throw new Error("dynamicUserList is not properly defined or empty.");
+  }
+};
+
+/**
+ * This method is mostly same as the one in UserRepository but it includes a lot more relations which are specific requirement here
+ * TODO: Figure out how to keep it in UserRepository and use it here
+ */
+export const findUsersByUsername = async ({
```
Comment (Udit-takkar) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:71]: no change in the code here.

Comment (Udit-takkar) [reply packages/features/bookings/lib/handleNewBooking/loadUsers.ts:71]: this function returns few more fields(credentials and metadata) than the static function in UserRepository.

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  return users.length ? users : eventType.users;
+};
+
+const validateHosts = (hosts: Hosts): void => {
+  if (!Array.isArray(hosts)) {
+    throw new Error("eventType.hosts is not properly defined.");
+  }
+};
+
+const loadDynamicUsers = async (dynamicUserList: string[], req: IncomingMessage) => {
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:52]: I think we could avoid sending the entire `req` object to this function. It needs just `orgSlug`. We can compute orgSlug and pass that directly to it.

**Benefits**
- Makes it easy to unit test as arguments are simpler
- Plus, it is easier to understand and provide it what this function needs.

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  validateDynamicUserList(dynamicUserList);
+  const { isValidOrgDomain, currentOrgDomain } = orgDomainConfig(req);
+  return findUsersByUsername({
+    usernameList: dynamicUserList,
+    orgSlug: isValidOrgDomain ? currentOrgDomain : null,
+  });
+};
+
+const validateDynamicUserList = (dynamicUserList: string[]): void => {
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:61]: It isn't doing anything dynamic user related. How about calling it something like `ensureArrayWithAtleastOneItem`? It makes it more usable.

Also, `dynamicUserList` is already being enforced as an array in TypeScript so, we just need to check that it has atleast one entry. Maybe that can be done inline in the calling function itself.

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import prisma, { userSelect } from "@calcom/prisma";
+import { credentialForCalendarServiceSelect } from "@calcom/prisma/selects/credential";
+
+import type { NewBookingEventType } from "../handleNewBooking";
+
+const log = logger.getSubLogger({ prefix: ["[loadUsers]:handleNewBooking "] });
+type Hosts = NewBookingEventType["hosts"];
+
+export const loadUsers = async (
+  eventType: NewBookingEventType,
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:17]: nit: I think we could simplify this type to pick or have just `hosts` and `users`

**Review by hariombalhara — DISMISSED** [review]
@Udit-takkar Left a few suggestions for code improvement

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  const users = hosts.map(({ user, isFixed, priority }) => ({
+    ...user,
+    isFixed,
+    priority,
+  }));
+  return users.length ? users : eventType.users;
+};
+
+const validateHosts = (hosts: Hosts): void => {
+  if (!Array.isArray(hosts)) {
```
[the remark this answers is not in the document]
Comment (Udit-takkar) [reply packages/features/bookings/lib/handleNewBooking/loadUsers.ts:47]: Yeah i was thinking about it removing but didn't know why we had these checks in the handleNewBooking

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  const users = hosts.map(({ user, isFixed, priority }) => ({
+    ...user,
+    isFixed,
+    priority,
+  }));
+  return users.length ? users : eventType.users;
+};
+
+const loadDynamicUsers = async (dynamicUserList: string[], currentOrgDomain: string | null) => {
+  if (!Array.isArray(dynamicUserList) || dynamicUserList.length === 0) {
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:43]:
```suggestion
  if (dynamicUserList.length === 0) {
```

File: packages/features/bookings/lib/handleNewBooking/loadUsers.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    throw new Error("dynamicUserList is not properly defined or empty.");
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking/loadUsers.ts:44]:
```suggestion
    throw new Error("dynamicUserList is empty.");
```

**Review by hariombalhara — APPROVED** [review]

**Pull request merged**